rclcpp/include/rclcpp/topic_statistics/subscriber_topic_statistics.hpp

```diff
@@ -0,0 +1,208 @@
+#include "libstatistics_collector/moving_average_statistics/types.hpp"
+
+namespace
+{
+/// Return the current nanoseconds (count) since epoch.
+/**
+ * \return the current nanoseconds (count) since epoch
+ */
+int64_t get_current_nanoseconds_since_epoch()
+{
+  const auto now = std::chrono::system_clock::now();
+  return std::chrono::duration_cast<std::chrono::nanoseconds>(now.time_since_epoch()).count();
+}
+}  // namespace
+
+namespace rclcpp
+{
+namespace topic_statistics
+{
+
+constexpr const char kDefaultPublishTopicName[]{"topic_statistics"};
+constexpr const std::chrono::milliseconds kDefaultPublishingPeriod{std::chrono::minutes(1)};
+
+using libstatistics_collector::collector::GenerateStatisticMessage;
+using metrics_statistics_msgs::msg::MetricsMessage;
+using libstatistics_collector::moving_average_statistics::StatisticData;
+
+/**
+ * Class used to collect, measure, and publish topic statistics data. Current statistics
+ * supported for subscribers are received message age and received message period.
+ *
+ * \tparam CallbackMessageT the subscribed message type
+ */
+template<typename CallbackMessageT>
+class SubscriberTopicStatistics
+{
+  using TopicStatsCollector =
+    libstatistics_collector::topic_statistics_collector::TopicStatisticsCollector<
+    CallbackMessageT>;
+  using ReceivedMessagePeriod =
+    libstatistics_collector::topic_statistics_collector::ReceivedMessagePeriodCollector<
+    CallbackMessageT>;
+
+public:
+  /// Construct a SubcriberTopicStatistics object.
+  /**
+  * This object wraps utilities, defined in libstatistics_collector, to collect,
+  * measure, and publish topic statistics data.
+   *
+  * \param node the node creating the subscription, used to create the publisher and timer to
+  * publish topic statistics.
+  * \param publishing_topic the topic to publish statistics to
+  * \param publishing_period the period at which topic statistics messages are published
+  */
+  SubscriberTopicStatistics(
+    rclcpp::Node & node,
+    const std::string & publishing_topic = kDefaultPublishTopicName,
+    const std::chrono::milliseconds & publishing_period = kDefaultPublishingPeriod)
+  {
+    publisher_ =
+      node.create_publisher<metrics_statistics_msgs::msg::MetricsMessage>(
+      publishing_topic,
+      10);
+
+    auto callback = [this]()
+      {
+        this->publish_message();
+      };
+
+    publisher_timer_ = node.create_wall_timer(publishing_period, callback);
+
+    node_name_ = node.get_name();
+
+    bring_up();
+  }
+
+  virtual ~SubscriberTopicStatistics()
+  {
+    tear_down();
+  }
+
+  /// Handle a message received by the subscription to collect statistics.
+  /**
+   * \param received_message the message received by the subscription
+   * \param now_nanoseconds current time in nanoseconds
+   */
+  virtual void handle_message(
+    const CallbackMessageT & received_message,
+    const rcl_time_point_value_t now_nanoseconds) const
```

Comment:
Since this is a public function, it might be nice to avoid the C types, and prefer something like `std::chrono::nanoseconds`.

Reply:
Would this be an issue if we were to use ros time in the future?

Reply:
Well `rcl_time_point_value_t` is implicitly a steady clock, whether or not it comes from ROS time or the system is an implementation detail. What is that time stamp used for? It seems like you would most often be interested in statistics in terms of actual wall time, but I don't know. Maybe you'd want both. Either way you could use `rclcpp::Time` if you want to potentially support ROS Time. My point here is that you should try to use a C++ construct in a C++ interface if possible.

Reply:
Sounds good, I've made the changes to use `rclcpp::Time`.